Fixes to context menus in wxDVC on macOS

vslavik opened the pull request:
The primary motivation for this PR was f370e9e8ea5bfbed8a6be0bd4e386ee2a3a4ddfe to fix https://github.com/vslavik/poedit/issues/849, and then I went down the rabbit hole a bit:

1. 37efaacba3dfdfe98520f5a2950ddb75a638b8cd — This is a no-brainer, prevents accidental menu clicks and makes wxOSX consistent with both other ports and native Cocoa context menus. The only question is whether to do it like this or limit oneself to the `wxDefaultPosition` case. I initially did the latter, but this seems more sensible and benefits legacy code that may pass explicit mouse coordinates.

2. f370e9e8ea5bfbed8a6be0bd4e386ee2a3a4ddfe — This fixes the weird behavior when using <kb>control</kbd>-click instead of right-click: there seems to be some corruption of internal state (AppKit, not wx) if `menuForEvent:` doesn't return a menu and we pop one up manually. (When I return a native menu, things behave normally. When I delay `PopupMenu` into next event loop iteration, i.e. after `menuForEvent:` returns, it doesn't help.) This code was there since the very beginning (~2009) and never meaningfully modified, so I don't think there was a particularly deep reason for using `menuForEvent:` [^1]

3. caffdd753a37cecd8be492157e49983a3c6b7f0a — While doing that, I noticed that the mouse position is never used... Apparently, `wxEVT_DATAVIEW_ITEM_CONTEXT_MENU` was always broken and didn't operate on the right-clicked item, but instead used whatever the current selection was. So I fixed that too. 

All of this should be cherry-picked into 3.2 too.

[^1]: In principle, it would be better to use `menuForEvent:` everywhere — see also https://github.com/wxWidgets/wxWidgets/issues/13010#issuecomment-1010846630 — but that would require a new API, e.g. addition of `virtual wxWindow::CreateContextMenu()` to be implementable.

vslavik committed:
Don't preselect popup menu items on macOS

Native macOS context menus are shown slightly to the right of the mouse
cursor so that no item is initially selected. Other platforms do the
same thing.

Adjust placement position manually to ensure the same behavior when
popping up the menu manually.

vslavik committed:
Don't use menuForEvent: in wxDataViewCtrl on macOS

Returning nil from menuForEvent: and then popping up a menu manually
causes some kind of Cocoa internal state confusion if the right-click
was performed by holding down Control and clicking (as is the Mac
alternative to right-click). See https://github.com/vslavik/poedit/issues/849

Implement wxEVT_DATAVIEW_ITEM_CONTEXT_MENU event using a handler for
the regular wxEVT_CONTEXT_MENU event instead.

vslavik committed:
Fix wxEVT_DATAVIEW_ITEM_CONTEXT_MENU item on macOS

Pass the item under mouse cursor to wxDataViewEvent.

Previously, first item of current selection was used instead. This was
inconsistent with all other ports and contrary to user expectation: the
UI was showing context menu on a different item from the one the action
was taken on.

csomor commented:
@vslavik thanks a lot, looking forward :-)

vadz reviewed (changes requested):
Thanks! I don't have any comments about this, all the explanations make sense to me.

I believe the original code was written by Hartwig, so I'm cc'ing @hwiesmann in case you have any comments. Otherwise I'll merge (and backport) this soon.

Thanks again!

hwiesmann commented:
No comments!

csomor commented on src/osx/dataview_osx.cpp:
I don't know about the expectations in the `wxEVT_DATAVIEW_ITEM_CONTEXT_MENU` Handler, but perhaps using the `wxDataViewEvent` constructor that also takes the `wxDataViewColumn` would allow to differentiate on the column as well if wanted?

csomor reviewed (approved):
I'm fine with it, added one possible change about the additional `wxDataViewColumn` info, this wasn't there before either in this implementation, but eg in the generic implementation it is filled when this event occurs.

vslavik committed:
Set column for wxEVT_DATAVIEW_ITEM_CONTEXT_MENU in all ports

Set wxDataViewEvent's column in both wxOSX and wxGTK implementations.
Generic implementation already did it.

vslavik commented:
> I'm fine with it, added one possible change about the additional wxDataViewColumn info,

Thanks!

That's trivial to do and worth doing. I pushed cd5150f2 that fixes this for both wxOSX and wxGTK.

csomor commented:
> > I'm fine with it, added one possible change about the additional wxDataViewColumn info,
> 
> Thanks!
> 
> That's trivial to do and worth doing. I pushed [42ecd8b](https://github.com/wxWidgets/wxWidgets/commit/42ecd8be4274501d3f7b7f14363a185797766b6a) that fixes this for both wxOSX and wxGTK.

perfect, thanks a lot, 

so @vadz please go ahead and merge when it's convenient for you
